### ISSUE TITLE
Polars ignoring rows that are empty in Excel

### ISSUE THREAD

**Issue by Rashik-raj** [pola-rs/polars#18250]
### Checks

- [X] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
import pandas as pd

def read_excel_using_polars(path):
    return pl.read_excel(source=path)

def read_csv_using_polars(path):
    return pl.read_csv(source=path)

def read_excel_using_pandas(path):
    return pd.read_excel(io=path)

def read_csv_using_pandas(path):
    return pd.read_csv(path)

pl_df_excel = read_excel_using_polars("test_empty_rows.xlsx")
pd_df_excel =read_excel_using_pandas("test_empty_rows.xlsx")

pl_df_csv = read_csv_using_polars("test_empty_rows.csv")
pd_df_csv = read_csv_using_pandas("test_empty_rows.csv")

print("Reading excel from polars", end="\n\n")
print(pl_df_excel)

print("\nReading excel from pandas", end="\n\n")
print(pd_df_excel)

print("\nReading csv from polars", end="\n\n")
print(pl_df_csv)

print("\nReading csv from pandas", end="\n\n")
print(pd_df_csv)
```
[test_empty_rows.xlsx](https://github.com/user-attachments/files/16647608/test_empty_rows.xlsx)
[test_empty_rows.csv](https://github.com/user-attachments/files/16647609/test_empty_rows.csv)

### Log output

```shell
Reading excel from polars

shape: (8, 4)
┌──────┬──────┬──────┬─────┐
│ a    ┆ b    ┆ c    ┆ d   │
│ ---  ┆ ---  ┆ ---  ┆ --- │
│ i64  ┆ i64  ┆ null ┆ i64 │
╞══════╪══════╪══════╪═════╡
│ 1    ┆ 1    ┆ null ┆ 1   │
│ 2    ┆ null ┆ null ┆ 2   │
│ 3    ┆ 3    ┆ null ┆ 3   │
│ null ┆ 4    ┆ null ┆ 4   │
│ 5    ┆ 5    ┆ null ┆ 5   │
│ 6    ┆ 6    ┆ null ┆ 6   │
│ 9    ┆ 9    ┆ null ┆ 9   │
│ 10   ┆ 10   ┆ null ┆ 0   │
└──────┴──────┴──────┴─────┘

Reading excel from pandas

      a     b   c    d
0   1.0   1.0 NaN  1.0
1   2.0   NaN NaN  2.0
2   3.0   3.0 NaN  3.0
3   NaN   4.0 NaN  4.0
4   5.0   5.0 NaN  5.0
5   6.0   6.0 NaN  6.0
6   NaN   NaN NaN  NaN
7   NaN   NaN NaN  NaN
8   9.0   9.0 NaN  9.0
9  10.0  10.0 NaN  0.0

Reading csv from polars

shape: (10, 4)
┌──────┬──────┬──────┬──────┐
│ a    ┆ b    ┆ c    ┆ d    │
│ ---  ┆ ---  ┆ ---  ┆ ---  │
│ i64  ┆ i64  ┆ str  ┆ i64  │
╞══════╪══════╪══════╪══════╡
│ 1    ┆ 1    ┆ null ┆ 1    │
│ 2    ┆ null ┆ null ┆ 2    │
│ 3    ┆ 3    ┆ null ┆ 3    │
│ null ┆ 4    ┆ null ┆ 4    │
│ 5    ┆ 5    ┆ null ┆ 5    │
│ 6    ┆ 6    ┆ null ┆ 6    │
│ null ┆ null ┆ null ┆ null │
│ null ┆ null ┆ null ┆ null │
│ 9    ┆ 9    ┆ null ┆ 9    │
│ 10   ┆ 10   ┆ null ┆ 0    │
└──────┴──────┴──────┴──────┘

Reading csv from pandas

      a     b   c    d
0   1.0   1.0 NaN  1.0
1   2.0   NaN NaN  2.0
2   3.0   3.0 NaN  3.0
3   NaN   4.0 NaN  4.0
4   5.0   5.0 NaN  5.0
5   6.0   6.0 NaN  6.0
6   NaN   NaN NaN  NaN
7   NaN   NaN NaN  NaN
8   9.0   9.0 NaN  9.0
9  10.0  10.0 NaN  0.0
```

### Issue description

We were using pandas to import files in our FastApi server that imports Excel, CSV, or XPT files. Due to lack of performance, we switched to Polars (had issue running in uvicorn as it is thread based server, made a workaround). We noticed that Polars is dropping rows that are entirely empty for Excel files.

To double check, I tested with pandas as well, and ensured that pandas do not drop empty rows. I went through polars documentation, but with no luck. performed some debugging in polars package and found `_drop_null_data` to be the culprit. I had monkeypatched `_drop_null_data` to return entire dataframe in an attempt to not loose empty rows as a workaround.

```python
import polars as pl

def _custom_drop_null_data(df, raise_if_empty):
    return df

pl.io.spreadsheet.functions._drop_null_data = _custom_drop_null_data
df = pl.read_excel(source="test_empty_rows.xlsx")  # we get empty rows as well
```

Also polars seems to be dropping empty rows only for Excel files which you can see from the reproducible sample. So to keep backward compatibility just in case it is desired feature, I am proposing to use `drop_empty_rows` parameter in `_drop_null_data` that defaults to `True`, and can be switched when reading Excel files as `pl.read_excel(source=path, drop_empty_rows=False)`. I have also verified that this happens with every engine of Excel.

This is how patched `_drop_null_data` will look like:
```python
def _drop_null_data(df: pl.DataFrame, *, raise_if_empty: bool, drop_empty_rows: bool = True) -> pl.DataFrame:
    """
    If DataFrame contains columns/rows that contain only nulls, drop them.

    If `drop_empty_rows` is set to `False`, empty rows are not dropped.
    """
    null_cols = []
    for col_name in df.columns:
        # note that if multiple unnamed columns are found then all but the first one
        # will be named as "_duplicated_{n}" (or "__UNNAMED__{n}" from calamine)
        if col_name == "" or re.match(r"(_duplicated_|__UNNAMED__)\d+$", col_name):
            col = df[col_name]
            if (
                col.dtype == Null
                or col.null_count() == len(df)
                or (
                    col.dtype in NUMERIC_DTYPES
                    and col.replace(0, None).null_count() == len(df)
                )
            ):
                null_cols.append(col_name)
    if null_cols:
        df = df.drop(*null_cols)

    if len(df) == 0 and len(df.columns) == 0:
        return _empty_frame(raise_if_empty)
    if drop_empty_rows:
        return df.filter(~F.all_horizontal(F.all().is_null()))
    return df
```

### Expected behavior

To not drop empty rows conditionally based on new flag `drop_empty_rows`.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.5.0
Index type:          UInt32
Platform:            macOS-14.6.1-arm64-arm-64bit
Python:              3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)]

----Optional dependencies----
adbc_driver_manager  1.1.0
cloudpickle          3.0.0
connectorx           0.3.3
deltalake            0.19.0
fastexcel            0.11.5
fsspec               2024.6.1
gevent               24.2.1
great_tables         0.10.0
hvplot               0.10.0
matplotlib           3.9.2
nest_asyncio         1.6.0
numpy                2.0.1
openpyxl             3.1.5
pandas               2.2.2
pyarrow              17.0.0
pydantic             2.8.2
pyiceberg            <not installed>
sqlalchemy           2.0.32
torch                <not installed>
xlsx2csv             0.8.3
xlsxwriter           3.2.0
```

</details>

**Comment by Rashik-raj** [comment]
Found out that there is open ticket [14874](https://github.com/pola-rs/polars/issues/14874). Somehow I did not find this 2 week back.

**Comment by doko123** [comment]
I think there is still issue when having file with leading empty rows:
Example:
Null null null
Null null null
A.     B.     C
1.      2.      3
4.      5.      6

when read_excel with drop empty rows and header_row=2 it reads the headers 4,5,6 instead of A, B, C.

unless I’m missing sth?

**Comment by cmdlineluser** [comment]
@doko123 Do you have an example?

I cannot seem to replicate that locally:

```python
import polars as pl

pl.DataFrame({
    "A": [None, None, "A", "1", "4"],
    "B": [None, None, "B", "2", "5"],
    "C": [None, None, "C", "3", "6"],
}).write_excel("18250.xlsx", include_header=False)
```

```python
__import__("fastexcel").__version__
# '0.12.0'
```

```
pl.read_excel("18250.xlsx")
# shape: (2, 3)
# ┌─────┬─────┬─────┐
# │ A   ┆ B   ┆ C   │
# │ --- ┆ --- ┆ --- │
# │ str ┆ str ┆ str │
# ╞═════╪═════╪═════╡
# │ 1   ┆ 2   ┆ 3   │
# │ 4   ┆ 5   ┆ 6   │
# └─────┴─────┴─────┘
```

```
pl.read_excel("18250.xlsx", read_options={"header_row": 2})
# shape: (2, 3)
# ┌─────┬─────┬─────┐
# │ A   ┆ B   ┆ C   │
# │ --- ┆ --- ┆ --- │
# │ str ┆ str ┆ str │
# ╞═════╪═════╪═════╡
# │ 1   ┆ 2   ┆ 3   │
# │ 4   ┆ 5   ┆ 6   │
# └─────┴─────┴─────┘
```

**Comment by doko123** [comment]
Maybe I will explain the scenario:
Given excel that have sometimes empty 2 rows and starting table on third row and sometimes have metadata (Applied filters…) on first line and second stays empty. 
min that case when reading excel and passing header_row=2 
For second scenario it reads columns correctly.
For first scenario: having empty rows it does not read columns correctly, as with header_row=2 counted from first non empty line in excel. So with attribute drop_empty_rows=False I was expecting that no matter if first two lines of excel are filled or not, the header_row is always taken from third line of excel, not third row of non empty rows.

but maybe there is another option I should use?

**Comment by doko123** [comment]
Im not at computer right now, will try to add an excel file for that example that have empty first two lines of file and starting real data to load on third row.

**Comment by doko123** [comment]
Hi.

Im sorry for delay. 

The issue is:

Having excel to read with always headers defined in 3rd row (for polars its header_row=2).
Sometimes the excel has metadata left on first line (Applied filters) that are treated as value of row.

While loading the excel which content is the same, but with applied firrent filters (none or some). The polars treats it differentially. Pandas for example ignores metadata actually.

Example of the same content of excel file:
> With metadata (applied filters)
![image](https://github.com/user-attachments/assets/3ef2a5fd-e83b-4e79-a98f-cd327a1e2987)
Without metadata (no filters)
![image](https://github.com/user-attachments/assets/60534e39-03f4-4ab7-bc33-edee55c1ae5c)

the same code for the same excel but with applied filters fails, and for no filters applied- passes:

```
    import polars as pl

    df = pl.read_excel(
        os.path.join(Path(__file__).parent, "example.xlsx"), read_options={"header_row": 2}, drop_empty_rows=False,
        columns=["A", "B", "C"]
    )

```

So I understood, to ignore empty rows from beginning of excel file if any, we can use drop_empty_rows=False. But it does not work, as starts counting from first NON-EMPTY row:
```
E       _fastexcel.ColumnNotFoundError: column with name "A" not found
E       Context:
E           0: available columns are: [ColumnInfo { name: "4", index: 0, dtype: Float, column_name_from: LookedUp, dtype_from: Guessed }, ColumnInfo { name: "5", index: 1, dtype: Float, column_name_from: LookedUp, dtype_from: Guessed }, ColumnInfo { name: "6", index: 2, dtype: Float, column_name_from: LookedUp, dtype_from: Guessed }]
```

**Comment by doko123** [comment]
Sorry edited: 
if metadata added (filters applied) the polars founds columns in 3rd row ( 2nd when countgng from 0)
if metadata not added (empty first row), polars thinks the 5th row is a header_row as it starts counting from first non empty row (3rd + header row 2)

**Comment by cmdlineluser** [comment]
Do you have fastexcel==0.11.6 installed?

I can replicate that behaviour on 0.11.6 but it is "fixed" on 0.12.0
